File: 39-combination-sum/39-combination-sum.java
class Solution {
    
    public void findCombo(int idx, int target, int[] candidates, List<List<Integer>> ans, List<Integer> ds)
    {
        if(idx == candidates.length)
        {
            if(target == 0)
                ans.add(new  ArrayList<>(ds));
            
            return;
        }
        
        // take
        if(candidates[idx] <= target)
        {
            ds.add(candidates[idx]);
            findCombo(idx, target - candidates[idx], candidates, ans, ds);
            ds.remove(ds.size() - 1);
        }
        
        // not take
        findCombo(idx + 1, target, candidates, ans, ds);
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        List<List<Integer>> ans = new  ArrayList<>();
        List<Integer> ds = new  ArrayList<>();
        findCombo(0, target, candidates, ans, ds);
        return ans;
    }
}

File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    
    void backtrack(int ind, int target, vector<int> &candidates, vector<int> &arr, vector<vector<int>> &res) {
        
        if(ind == candidates.size()) {
            if(target == 0) res.push_back(arr);
            return;
        }
        
        if(candidates[ind] <= target) {
            // pick
            arr.push_back(candidates[ind]);
            backtrack(ind, target - candidates[ind], candidates, arr, res);
            arr.pop_back();
        }

        // not pick
        backtrack(ind + 1, target, candidates, arr, res);
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> arr;
        vector<vector<int>> res;
        backtrack(0, target, candidates, arr, res);
        return res;
    }
};